Test suite

jausions opened the pull request:
A test suite runner and run comparator that is written in PHP for portability over different operating systems.

`@group` annotations have been added to the example files to be able to target specific ones to run.

I needed this script to run the tests under Windows to fix an issue with TCPDF under PHP8,1 (this will be a separate PR)

Tested on Windows 10 with the following versions of PHP:

- 5.3.29
- 5.4.44
- 5.5.38
- 5.6.40
- 7.0.10
- 7.1.15
- 7.2.28
- 7.3.15
- 7.4.33
- 8.0.25
- 8.1.12

CLAassistant commented:
[![CLA assistant check](https://cla-assistant.io/pull/badge/not_signed)](https://cla-assistant.io/tecnickcom/TCPDF?pullRequest=566) <br/>Thank you for your submission! We really appreciate it. Like many open source projects, we ask that you sign our [Contributor License Agreement](https://cla-assistant.io/tecnickcom/TCPDF?pullRequest=566) before we can accept your contribution.<br/><sub>You have signed the CLA already but the status is still pending? Let us [recheck](https://cla-assistant.io/check/tecnickcom/TCPDF?pullRequest=566) it.</sub>

williamdes commented on composer.json:
```diff
@@ -24,6 +24,9 @@
         "php": ">=5.3.0"
     },
+    "require-dev": {
+        "cs278/mktemp": "*"
```
Could you remove this and add it on the fly on the test script?

jausions replied:
@williamdes Done.

jausions commented:
For information, these are examples of output.

PHP 5.3 with `--output-dir` and `--group=cell` options:
===

```
pdfinfo: C:\Program Files\Xpdf\bin64\pdfinfo.exe
pdfinfo version: pdfinfo version 4.04 [www.xpdfreader.com]
Copyright 1996-2022 Glyph & Cog, LLC

PHP: "php5.3"
PHP version: PHP 5.3.29 (cli) (built: Aug 15 2014 19:03:18)
Copyright (c) 1997-2014 The PHP Group
Zend Engine v2.3.0, Copyright (c) 1998-2014 Zend Technologies

PHP extension folder: C:/PROGRA~2/PHP53~1.29(/ext
Extensions:
    bcmath: BUILT-IN
    gd: AVAILABLE
    imagick: NO
    json: BUILT-IN
    openssl: AVAILABLE
    pcov: NO
    xdebug: NO
    xml: BUILT-IN

Root folder: E:\Users\jausions\Documents\Projects\github\jausions\TCPDF\
Output folder: E:\Users\jausions\Documents\Projects\github\jausions\TCPDF\tests\runs\PHP5.3

Test folder: E:\Users\jausions\Documents\Projects\github\jausions\TCPDF\tests\
File: example_004.php
    Lint: PASSED
    Run: PASSED
    Output: PDF
    Output file: E:\Users\jausions\Documents\Projects\github\jausions\TCPDF\tests\runs\PHP5.3\example_004.output.pdf
File: example_005.php
    Lint: PASSED
    Run: PASSED
    Output: PDF
    Output file: E:\Users\jausions\Documents\Projects\github\jausions\TCPDF\tests\runs\PHP5.3\example_005.output.pdf
File: example_007.php
    Lint: PASSED
    Run: PASSED
    Output: PDF
    Output file: E:\Users\jausions\Documents\Projects\github\jausions\TCPDF\tests\runs\PHP5.3\example_007.output.pdf
File: example_017.php
    Lint: PASSED
    Run: PASSED
    Output: PDF
    Output file: E:\Users\jausions\Documents\Projects\github\jausions\TCPDF\tests\runs\PHP5.3\example_017.output.pdf
File: example_020.php
    Lint: PASSED
    Run: PASSED
    Output: PDF
    Output file: E:\Users\jausions\Documents\Projects\github\jausions\TCPDF\tests\runs\PHP5.3\example_020.output.pdf
File: example_034.php
    Lint: PASSED
    Run: PASSED
    Output: PDF
    Output file: E:\Users\jausions\Documents\Projects\github\jausions\TCPDF\tests\runs\PHP5.3\example_034.output.pdf
File: example_035.php
    Lint: PASSED
    Run: PASSED
    Output: PDF
    Output file: E:\Users\jausions\Documents\Projects\github\jausions\TCPDF\tests\runs\PHP5.3\example_035.output.pdf
File: example_057.php
    Lint: PASSED
    Run: PASSED
    Output: PDF
    Output file: E:\Users\jausions\Documents\Projects\github\jausions\TCPDF\tests\runs\PHP5.3\example_057.output.pdf
Generated files remaining on disk: 16
Test suite: PASSED
    Runtime: 5s
    Total tests: 8
Exit code: 0
```

PHP 8,1:
===
```
pdfinfo: C:\Program Files\Xpdf\bin64\pdfinfo.exe
pdfinfo version: pdfinfo version 4.04 [www.xpdfreader.com]
Copyright 1996-2022 Glyph & Cog, LLC

PHP: "C:\Program Files\PHP 8.1.12 (NTS)\php.exe"
PHP version: PHP 8.1.12 (cli) (built: Oct 25 2022 18:20:48) (NTS Visual C++ 2019 x64)
Copyright (c) The PHP Group
Zend Engine v4.1.12, Copyright (c) Zend Technologies

PHP extension folder: C:/PROGRA~1/PHP81~1.12(/ext
Extensions:
    bcmath: BUILT-IN
    gd: AVAILABLE
    imagick: NO
    json: BUILT-IN
    openssl: AVAILABLE
    pcov: NO
    xdebug: NO
    xml: BUILT-IN

Root folder: E:\Users\jausions\Documents\Projects\github\jausions\TCPDF\

The --output-dir option was not used, a temporary folder will be necessary.
Output folder: C:\Users\jausions\AppData\Local\Temp\TCPDF-tests.vWIN2N9Vt

Test folder: E:\Users\jausions\Documents\Projects\github\jausions\TCPDF\tests\
File: example_001.php
    Lint: PASSED
    Run: PASSED
    Output: PDF
File: example_002.php
    Lint: PASSED
    Run: PASSED
    Output: PDF
File: example_003.php
    Lint: PASSED
    Run: PASSED
    Output: PDF
File: example_004.php
    Lint: PASSED
    Run: PASSED
    Output: PDF
File: example_005.php
    Lint: PASSED
    Run: PASSED
    Output: PDF
File: example_006.php
    Run: SKIPPED
File: example_007.php
    Lint: PASSED
    Run: PASSED
    Output: PDF
File: example_008.php
    Lint: PASSED
    Run: PASSED
    Output: PDF
File: example_009.php
    Lint: PASSED
    Run: PASSED
    Output: PDF
File: example_010.php
    Lint: PASSED
    Run: PASSED
    Output: PDF
File: example_011.php
    Lint: PASSED
    Run: PASSED
    Output: PDF
File: example_012.php
    Lint: PASSED
    Run: PASSED
    Output: PDF
File: example_013.php
    Lint: PASSED
    Run: PASSED
    Output: PDF
File: example_014.php
    Lint: PASSED
    Run: PASSED
    Output: PDF
File: example_015.php
    Lint: PASSED
    Run: PASSED
    Output: PDF
File: example_016.php
    Lint: PASSED
    Run: PASSED
    Output: PDF
File: example_017.php
    Lint: PASSED
    Run: PASSED
    Output: PDF
File: example_018.php
    Lint: PASSED
    Run: PASSED
    Output: PDF
File: example_019.php
    Lint: PASSED
    Run: PASSED
    Output: PDF
File: example_020.php
    Lint: PASSED
    Run: PASSED
    Output: PDF
File: example_021.php
    Lint: PASSED
    Run: PASSED
    Output: PDF
File: example_022.php
    Lint: PASSED
    Run: PASSED
    Output: PDF
File: example_023.php
    Lint: PASSED
    Run: PASSED
    Output: PDF
File: example_024.php
    Lint: PASSED
    Run: PASSED
    Output: PDF
File: example_025.php
    Lint: PASSED
    Run: PASSED
    Output: PDF
File: example_026.php
    Lint: PASSED
    Run: PASSED
    Output: PDF
File: example_027.php
    Lint: PASSED
    Run: PASSED
    Output: PDF
File: example_028.php
    Lint: PASSED
    Run: PASSED
    Output: PDF
File: example_029.php
    Lint: PASSED
    Run: PASSED
    Output: PDF
File: example_030.php
    Lint: PASSED
    Run: PASSED
    Output: PDF
File: example_031.php
    Lint: PASSED
    Run: PASSED
    Output: PDF
File: example_032.php
    Lint: PASSED
    Run: PASSED
    Output: PDF
File: example_033.php
    Lint: PASSED
    Run: PASSED
    Output: PDF
File: example_034.php
    Lint: PASSED
    Run: PASSED
    Output: PDF
File: example_035.php
    Lint: PASSED
    Run: PASSED
    Output: PDF
File: example_036.php
    Lint: PASSED
    Run: PASSED
    Output: PDF
File: example_037.php
    Lint: PASSED
    Run: PASSED
    Output: PDF
File: example_038.php
    Lint: PASSED
    Run: PASSED
    Output: PDF
File: example_039.php
    Lint: PASSED
    Run: PASSED
    Output: PDF
File: example_040.php
    Lint: PASSED
    Run: PASSED
    Output: PDF
File: example_041.php
    Lint: PASSED
    Run: PASSED
    Output: PDF
File: example_042.php
    Lint: PASSED
    Run: PASSED
    Output: PDF
File: example_043.php
    Lint: PASSED
    Run: PASSED
    Output: PDF
File: example_044.php
    Lint: PASSED
    Run: PASSED
    Output: PDF
File: example_045.php
    Lint: PASSED
    Run: PASSED
    Output: PDF
File: example_046.php
    Lint: PASSED
    Run: PASSED
    Output: PDF
File: example_047.php
    Lint: PASSED
    Run: PASSED
    Output: PDF
File: example_048.php
    Lint: PASSED
    Run: PASSED
    Output: PDF
File: example_049.php
    Lint: PASSED
    Run: PASSED
    Output: PDF
File: example_050.php
    Lint: PASSED
    Run: PASSED
    Output: PDF
File: example_051.php
    Lint: PASSED
    Run: PASSED
    Output: PDF
File: example_052.php
    Lint: PASSED
    Run: PASSED
    Output: PDF
File: example_053.php
    Lint: PASSED
    Run: PASSED
    Output: PDF
File: example_054.php
    Lint: PASSED
    Run: PASSED
    Output: PDF
File: example_055.php
    Lint: PASSED
    Run: PASSED
    Output: PDF
File: example_056.php
    Lint: PASSED
    Run: PASSED
    Output: PDF
File: example_057.php
    Lint: PASSED
    Run: PASSED
    Output: PDF
File: example_058.php
    Lint: PASSED
    Run: PASSED
    Output: PDF
File: example_059.php
    Lint: PASSED
    Run: PASSED
    Output: PDF
File: example_060.php
    Lint: PASSED
    Run: PASSED
    Output: PDF
File: example_061.php
    Lint: PASSED
    Run: PASSED
    Output: PDF
File: example_062.php
    Lint: PASSED
    Run: PASSED
    Output: PDF
File: example_063.php
    Lint: PASSED
    Run: PASSED
    Output: PDF
File: example_064.php
    Lint: PASSED
    Run: PASSED
    Output: PDF
File: example_065.php
    Lint: PASSED
    Run: PASSED
    Output: PDF
File: example_066.php
    Lint: PASSED
    Run: PASSED
    Output: PDF
File: example_067.php
    Lint: PASSED
    Run: PASSED
    Output: PDF
File: barcodes/example_1d_html.php
    Lint: PASSED
    Run: PASSED
    Output: HTML
File: barcodes/example_1d_png.php
    Lint: PASSED
    Run: PASSED
    Output: PNG
File: barcodes/example_1d_svg.php
    Lint: PASSED
    Run: PASSED
    Output: SVG
File: barcodes/example_1d_svgi.php
    Lint: PASSED
    Run: PASSED
    Output: SVG
File: barcodes/example_2d_datamatrix_html.php
    Lint: PASSED
    Run: PASSED
    Output: HTML
File: barcodes/example_2d_datamatrix_png.php
    Lint: PASSED
    Run: PASSED
    Output: PNG
File: barcodes/example_2d_datamatrix_svg.php
    Lint: PASSED
    Run: PASSED
    Output: SVG
File: barcodes/example_2d_datamatrix_svgi.php
    Lint: PASSED
    Run: PASSED
    Output: SVG
File: barcodes/example_2d_pdf417_html.php
    Lint: PASSED
    Run: PASSED
    Output: HTML
File: barcodes/example_2d_pdf417_png.php
    Lint: PASSED
    Run: PASSED
    Output: PNG
File: barcodes/example_2d_pdf417_svg.php
    Lint: PASSED
    Run: PASSED
    Output: SVG
File: barcodes/example_2d_pdf417_svgi.php
    Lint: PASSED
    Run: PASSED
    Output: SVG
File: barcodes/example_2d_qrcode_html.php
    Lint: PASSED
    Run: PASSED
    Output: NOT HTML FILE
    Logs:
---------------------------

Deprecated: Implicit conversion from float 0.5 to int loses precision in E:\Users\jausions\Documents\Projects\github\jausions\TCPDF\include\barcodes\qrcode.php on line 891

Deprecated: Implicit conversion from float 1.5 to int loses precision in E:\Users\jausions\Documents\Projects\github\jausions\TCPDF\include\barcodes\qrcode.php on line 891

Deprecated: Implicit conversion from float 2.5 to int loses precision in E:\Users\jausions\Documents\Projects\github\jausions\TCPDF\include\barcodes\qrcode.php on line 891

Deprecated: Implicit conversion from float 3.5 to int loses precision in E:\Users\jausions\Documents\Projects\github\jausions\TCPDF\include\barcodes\qrcode.php on line 891

Deprecated: Implicit conversion from float 4.5 to int loses precision in E:\Users\jausions\Documents\Projects\github\jausions\TCPDF\include\barcodes\qrcode.php on line 891

Deprecated: Implicit conversion from float 5.5 to int loses precision in E:\Users\jausions\Documents\Projects\github\jausions\TCPDF\include\barcodes\qrcode.php on line 891

Deprecated: Implicit conversion from float 6.5 to int loses precision in E:\Users\jausions\Documents\Projects\github\jausions\TCPDF\include\barcodes\qrcode.php on line 891

Deprecated: Implicit conversion from float 7.5 to int loses precision in E:\Users\jausions\Documents\Projects\github\jausions\TCPDF\include\barcodes\qrcode.php on line 891

Deprecated: Implicit conversion from float 8.5 to int loses precision in E:\Users\jausions\Documents\Projects\github\jausions\TCPDF\include\barcodes\qrcode.php on line 891

Deprecated: Implicit conversion from float 9.5 to int loses precision in E:\Users\jausions\Documents\Projects\github\jausions\TCPDF\include\barcodes\qrcode.php on line 891

Deprecated: Implicit conversion from float 10.5 to int loses precision in E:\Users\jausions\Documents\Projects\github\jausions\TCPDF\include\barcodes\qrcode.php on line 891

Deprecated: Implicit conversion from float 11.5 to int loses precision in E:\Users\jausions\Documents\Projects\github\jausions\TCPDF\include\barcodes\qrcode.php on line 891

Deprecated: Implicit conversion from float 12.5 to int loses precision in E:\Users\jausions\Documents\Projects\github\jausions\TCPDF\include\barcodes\qrcode.php on line 891

Deprecated: Implicit conversion from float 0.5 to int loses precision in E:\Users\jausions\Documents\Projects\github\jausions\TCPDF\include\barcodes\qrcode.php on line 895

Deprecated: Implicit conversion from float 1.5 to int loses precision in E:\Users\jausions\Documents\Projects\github\jausions\TCPDF\include\barcodes\qrcode.php on line 895

Deprecated: Implicit conversion from float 2.5 to int loses precision in E:\Users\jausions\Documents\Projects\github\jausions\TCPDF\include\barcodes\qrcode.php on line 895

Deprecated: Implicit conversion from float 3.5 to int loses precision in E:\Users\jausions\Documents\Projects\github\jausions\TCPDF\include\barcodes\qrcode.php on line 895

Deprecated: Implicit conversion from float 4.5 to int loses precision in E:\Users\jausions\Documents\Projects\github\jausions\TCPDF\include\barcodes\qrcode.php on line 895

Deprecated: Implicit conversion from float 5.5 to int loses precision in E:\Users\jausions\Documents\Projects\github\jausions\TCPDF\include\barcodes\qrcode.php on line 895

Deprecated: Implicit conversion from float 6.5 to int loses precision in E:\Users\jausions\Documents\Projects\github\jausions\TCPDF\include\barcodes\qrcode.php on line 895

Deprecated: Implicit conversion from float 7.5 to int loses precision in E:\Users\jausions\Documents\Projects\github\jausions\TCPDF\include\barcodes\qrcode.php on line 895

Deprecated: Implicit conversion from float 8.5 to int loses precision in E:\Users\jausions\Documents\Projects\github\jausions\TCPDF\include\barcodes\qrcode.php on line 895

Deprecated: Implicit conversion from float 9.5 to int loses precision in E:\Users\jausions\Documents\Projects\github\jausions\TCPDF\include\barcodes\qrcode.php on line 895

Deprecated: Implicit conversion from float 10.5 to int loses precision in E:\Users\jausions\Documents\Projects\github\jausions\TCPDF\include\barcodes\qrcode.php on line 895

Deprecated: Implicit conversion from float 11.5 to int loses precision in E:\Users\jausions\Documents\Projects\github\jausions\TCPDF\include\barcodes\qrcode.php on line 895

Deprecated: Implicit conversion from float 12.5 to int loses precision in E:\Users\jausions\Documents\Projects\github\jausions\TCPDF\include\barcodes\qrcode.php on line 895

Deprecated: Implicit conversion from float 13.5 to int loses precision in E:\Users\jausions\Documents\Projects\github\jausions\TCPDF\include\barcodes\qrcode.php on line 895

Deprecated: Implicit conversion from float 14.5 to int loses precision in E:\Users\jausions\Documents\Projects\github\jausions\TCPDF\include\barcodes\qrcode.php on line 895

Deprecated: Implicit conversion from float 15.5 to int loses precision in E:\Users\jausions\Documents\Projects\github\jausions\TCPDF\include\barcodes\qrcode.php on line 895

Deprecated: Implicit conversion from float 16.5 to int loses precision in E:\Users\jausions\Documents\Projects\github\jausions\TCPDF\include\barcodes\qrcode.php on line 895

Deprecated: Implicit conversion from float 17.5 to int loses precision in E:\Users\jausions\Documents\Projects\github\jausions\TCPDF\include\barcodes\qrcode.php on line 895

Deprecated: Implicit conversion from float 18.5 to int loses precision in E:\Users\jausions\Documents\Projects\github\jausions\TCPDF\include\barcodes\qrcode.php on line 895

Deprecated: Implicit conversion from float 19.5 to int loses precision in E:\Users\jausions\Documents\Projects\github\jausions\TCPDF\include\barcodes\qrcode.php on line 895

Deprecated: Implicit conversion from float 20.5 to int loses precision in E:\Users\jausions\Documents\Projects\github\jausions\TCPDF\include\barcodes\qrcode.php on line 895

Deprecated: Implicit conversion from float 21.5 to int loses precision in E:\Users\jausions\Documents\Projects\github\jausions\TCPDF\include\barcodes\qrcode.php on line 895

---------------------------
File: barcodes/example_2d_qrcode_png.php
    Lint: PASSED
    Run: PASSED
    Output: NOT PNG FILE
    Logs:
---------------------------

Deprecated: Implicit conversion from float 0.5 to int loses precision in E:\Users\jausions\Documents\Projects\github\jausions\TCPDF\include\barcodes\qrcode.php on line 891

Deprecated: Implicit conversion from float 1.5 to int loses precision in E:\Users\jausions\Documents\Projects\github\jausions\TCPDF\include\barcodes\qrcode.php on line 891

Deprecated: Implicit conversion from float 2.5 to int loses precision in E:\Users\jausions\Documents\Projects\github\jausions\TCPDF\include\barcodes\qrcode.php on line 891

Deprecated: Implicit conversion from float 3.5 to int loses precision in E:\Users\jausions\Documents\Projects\github\jausions\TCPDF\include\barcodes\qrcode.php on line 891

Deprecated: Implicit conversion from float 4.5 to int loses precision in E:\Users\jausions\Documents\Projects\github\jausions\TCPDF\include\barcodes\qrcode.php on line 891

Deprecated: Implicit conversion from float 5.5 to int loses precision in E:\Users\jausions\Documents\Projects\github\jausions\TCPDF\include\barcodes\qrcode.php on line 891

Deprecated: Implicit conversion from float 6.5 to int loses precision in E:\Users\jausions\Documents\Projects\github\jausions\TCPDF\include\barcodes\qrcode.php on line 891

Deprecated: Implicit conversion from float 7.5 to int loses precision in E:\Users\jausions\Documents\Projects\github\jausions\TCPDF\include\barcodes\qrcode.php on line 891

Deprecated: Implicit conversion from float 8.5 to int loses precision in E:\Users\jausions\Documents\Projects\github\jausions\TCPDF\include\barcodes\qrcode.php on line 891

Deprecated: Implicit conversion from float 9.5 to int loses precision in E:\Users\jausions\Documents\Projects\github\jausions\TCPDF\include\barcodes\qrcode.php on line 891

Deprecated: Implicit conversion from float 10.5 to int loses precision in E:\Users\jausions\Documents\Projects\github\jausions\TCPDF\include\barcodes\qrcode.php on line 891

Deprecated: Implicit conversion from float 11.5 to int loses precision in E:\Users\jausions\Documents\Projects\github\jausions\TCPDF\include\barcodes\qrcode.php on line 891

Deprecated: Implicit conversion from float 12.5 to int loses precision in E:\Users\jausions\Documents\Projects\github\jausions\TCPDF\include\barcodes\qrcode.php on line 891

Deprecated: Implicit conversion from float 0.5 to int loses precision in E:\Users\jausions\Documents\Projects\github\jausions\TCPDF\include\barcodes\qrcode.php on line 895

Deprecated: Implicit conversion from float 1.5 to int loses precision in E:\Users\jausions\Documents\Projects\github\jausions\TCPDF\include\barcodes\qrcode.php on line 895

Deprecated: Implicit conversion from float 2.5 to int loses precision in E:\Users\jausions\Documents\Projects\github\jausions\TCPDF\include\barcodes\qrcode.php on line 895

Deprecated: Implicit conversion from float 3.5 to int loses precision in E:\Users\jausions\Documents\Projects\github\jausions\TCPDF\include\barcodes\qrcode.php on line 895

Deprecated: Implicit conversion from float 4.5 to int loses precision in E:\Users\jausions\Documents\Projects\github\jausions\TCPDF\include\barcodes\qrcode.php on line 895

Deprecated: Implicit conversion from float 5.5 to int loses precision in E:\Users\jausions\Documents\Projects\github\jausions\TCPDF\include\barcodes\qrcode.php on line 895

Deprecated: Implicit conversion from float 6.5 to int loses precision in E:\Users\jausions\Documents\Projects\github\jausions\TCPDF\include\barcodes\qrcode.php on line 895

Deprecated: Implicit conversion from float 7.5 to int loses precision in E:\Users\jausions\Documents\Projects\github\jausions\TCPDF\include\barcodes\qrcode.php on line 895

Deprecated: Implicit conversion from float 8.5 to int loses precision in E:\Users\jausions\Documents\Projects\github\jausions\TCPDF\include\barcodes\qrcode.php on line 895

Deprecated: Implicit conversion from float 9.5 to int loses precision in E:\Users\jausions\Documents\Projects\github\jausions\TCPDF\include\barcodes\qrcode.php on line 895

Deprecated: Implicit conversion from float 10.5 to int loses precision in E:\Users\jausions\Documents\Projects\github\jausions\TCPDF\include\barcodes\qrcode.php on line 895

Deprecated: Implicit conversion from float 11.5 to int loses precision in E:\Users\jausions\Documents\Projects\github\jausions\TCPDF\include\barcodes\qrcode.php on line 895

Deprecated: Implicit conversion from float 12.5 to int loses precision in E:\Users\jausions\Documents\Projects\github\jausions\TCPDF\include\barcodes\qrcode.php on line 895

Deprecated: Implicit conversion from float 13.5 to int loses precision in E:\Users\jausions\Documents\Projects\github\jausions\TCPDF\include\barcodes\qrcode.php on line 895

Deprecated: Implicit conversion from float 14.5 to int loses precision in E:\Users\jausions\Documents\Projects\github\jausions\TCPDF\include\barcodes\qrcode.php on line 895

Deprecated: Implicit conversion from float 15.5 to int loses precision in E:\Users\jausions\Documents\Projects\github\jausions\TCPDF\include\barcodes\qrcode.php on line 895

Deprecated: Implicit conversion from float 16.5 to int loses precision in E:\Users\jausions\Documents\Projects\github\jausions\TCPDF\include\barcodes\qrcode.php on line 895

Deprecated: Implicit conversion from float 17.5 to int loses precision in E:\Users\jausions\Documents\Projects\github\jausions\TCPDF\include\barcodes\qrcode.php on line 895

Deprecated: Implicit conversion from float 18.5 to int loses precision in E:\Users\jausions\Documents\Projects\github\jausions\TCPDF\include\barcodes\qrcode.php on line 895

Deprecated: Implicit conversion from float 19.5 to int loses precision in E:\Users\jausions\Documents\Projects\github\jausions\TCPDF\include\barcodes\qrcode.php on line 895

Deprecated: Implicit conversion from float 20.5 to int loses precision in E:\Users\jausions\Documents\Projects\github\jausions\TCPDF\include\barcodes\qrcode.php on line 895

Deprecated: Implicit conversion from float 21.5 to int loses precision in E:\Users\jausions\Documents\Projects\github\jausions\TCPDF\include\barcodes\qrcode.php on line 895

Warning: Cannot modify header information - headers already sent by (output started at E:\Users\jausions\Documents\Projects\github\jausions\TCPDF\include\barcodes\qrcode.php:891) in E:\Users\jausions\Documents\Projects\github\jausions\TCPDF\tcpdf_barcodes_2d.php on line 175

Warning: Cannot modify header information - headers already sent by (output started at E:\Users\jausions\Documents\Projects\github\jausions\TCPDF\include\barcodes\qrcode.php:891) in E:\Users\jausions\Documents\Projects\github\jausions\TCPDF\tcpdf_barcodes_2d.php on line 176

Warning: Cannot modify header information - headers already sent by (output started at E:\Users\jausions\Documents\Projects\github\jausions\TCPDF\include\barcodes\qrcode.php:891) in E:\Users\jausions\Documents\Projects\github\jausions\TCPDF\tcpdf_barcodes_2d.php on line 177

Warning: Cannot modify header information - headers already sent by (output started at E:\Users\jausions\Documents\Projects\github\jausions\TCPDF\include\barcodes\qrcode.php:891) in E:\Users\jausions\Documents\Projects\github\jausions\TCPDF\tcpdf_barcodes_2d.php on line 178

Warning: Cannot modify header information - headers already sent by (output started at E:\Users\jausions\Documents\Projects\github\jausions\TCPDF\include\barcodes\qrcode.php:891) in E:\Users\jausions\Documents\Projects\github\jausions\TCPDF\tcpdf_barcodes_2d.php on line 179

---------------------------
File: barcodes/example_2d_qrcode_svg.php
    Lint: PASSED
    Run: PASSED
    Output: NOT SVG FILE
    Logs:
---------------------------

Deprecated: Implicit conversion from float 0.5 to int loses precision in E:\Users\jausions\Documents\Projects\github\jausions\TCPDF\include\barcodes\qrcode.php on line 891

Deprecated: Implicit conversion from float 1.5 to int loses precision in E:\Users\jausions\Documents\Projects\github\jausions\TCPDF\include\barcodes\qrcode.php on line 891

Deprecated: Implicit conversion from float 2.5 to int loses precision in E:\Users\jausions\Documents\Projects\github\jausions\TCPDF\include\barcodes\qrcode.php on line 891

Deprecated: Implicit conversion from float 3.5 to int loses precision in E:\Users\jausions\Documents\Projects\github\jausions\TCPDF\include\barcodes\qrcode.php on line 891

Deprecated: Implicit conversion from float 4.5 to int loses precision in E:\Users\jausions\Documents\Projects\github\jausions\TCPDF\include\barcodes\qrcode.php on line 891

Deprecated: Implicit conversion from float 5.5 to int loses precision in E:\Users\jausions\Documents\Projects\github\jausions\TCPDF\include\barcodes\qrcode.php on line 891

Deprecated: Implicit conversion from float 6.5 to int loses precision in E:\Users\jausions\Documents\Projects\github\jausions\TCPDF\include\barcodes\qrcode.php on line 891

Deprecated: Implicit conversion from float 7.5 to int loses precision in E:\Users\jausions\Documents\Projects\github\jausions\TCPDF\include\barcodes\qrcode.php on line 891

Deprecated: Implicit conversion from float 8.5 to int loses precision in E:\Users\jausions\Documents\Projects\github\jausions\TCPDF\include\barcodes\qrcode.php on line 891

Deprecated: Implicit conversion from float 9.5 to int loses precision in E:\Users\jausions\Documents\Projects\github\jausions\TCPDF\include\barcodes\qrcode.php on line 891

Deprecated: Implicit conversion from float 10.5 to int loses precision in E:\Users\jausions\Documents\Projects\github\jausions\TCPDF\include\barcodes\qrcode.php on line 891

Deprecated: Implicit conversion from float 11.5 to int loses precision in E:\Users\jausions\Documents\Projects\github\jausions\TCPDF\include\barcodes\qrcode.php on line 891

Deprecated: Implicit conversion from float 12.5 to int loses precision in E:\Users\jausions\Documents\Projects\github\jausions\TCPDF\include\barcodes\qrcode.php on line 891

Deprecated: Implicit conversion from float 0.5 to int loses precision in E:\Users\jausions\Documents\Projects\github\jausions\TCPDF\include\barcodes\qrcode.php on line 895

Deprecated: Implicit conversion from float 1.5 to int loses precision in E:\Users\jausions\Documents\Projects\github\jausions\TCPDF\include\barcodes\qrcode.php on line 895

Deprecated: Implicit conversion from float 2.5 to int loses precision in E:\Users\jausions\Documents\Projects\github\jausions\TCPDF\include\barcodes\qrcode.php on line 895

Deprecated: Implicit conversion from float 3.5 to int loses precision in E:\Users\jausions\Documents\Projects\github\jausions\TCPDF\include\barcodes\qrcode.php on line 895

Deprecated: Implicit conversion from float 4.5 to int loses precision in E:\Users\jausions\Documents\Projects\github\jausions\TCPDF\include\barcodes\qrcode.php on line 895

Deprecated: Implicit conversion from float 5.5 to int loses precision in E:\Users\jausions\Documents\Projects\github\jausions\TCPDF\include\barcodes\qrcode.php on line 895

Deprecated: Implicit conversion from float 6.5 to int loses precision in E:\Users\jausions\Documents\Projects\github\jausions\TCPDF\include\barcodes\qrcode.php on line 895

Deprecated: Implicit conversion from float 7.5 to int loses precision in E:\Users\jausions\Documents\Projects\github\jausions\TCPDF\include\barcodes\qrcode.php on line 895

Deprecated: Implicit conversion from float 8.5 to int loses precision in E:\Users\jausions\Documents\Projects\github\jausions\TCPDF\include\barcodes\qrcode.php on line 895

Deprecated: Implicit conversion from float 9.5 to int loses precision in E:\Users\jausions\Documents\Projects\github\jausions\TCPDF\include\barcodes\qrcode.php on line 895

Deprecated: Implicit conversion from float 10.5 to int loses precision in E:\Users\jausions\Documents\Projects\github\jausions\TCPDF\include\barcodes\qrcode.php on line 895

Deprecated: Implicit conversion from float 11.5 to int loses precision in E:\Users\jausions\Documents\Projects\github\jausions\TCPDF\include\barcodes\qrcode.php on line 895

Deprecated: Implicit conversion from float 12.5 to int loses precision in E:\Users\jausions\Documents\Projects\github\jausions\TCPDF\include\barcodes\qrcode.php on line 895

Deprecated: Implicit conversion from float 13.5 to int loses precision in E:\Users\jausions\Documents\Projects\github\jausions\TCPDF\include\barcodes\qrcode.php on line 895

Deprecated: Implicit conversion from float 14.5 to int loses precision in E:\Users\jausions\Documents\Projects\github\jausions\TCPDF\include\barcodes\qrcode.php on line 895

Deprecated: Implicit conversion from float 15.5 to int loses precision in E:\Users\jausions\Documents\Projects\github\jausions\TCPDF\include\barcodes\qrcode.php on line 895

Deprecated: Implicit conversion from float 16.5 to int loses precision in E:\Users\jausions\Documents\Projects\github\jausions\TCPDF\include\barcodes\qrcode.php on line 895

Deprecated: Implicit conversion from float 17.5 to int loses precision in E:\Users\jausions\Documents\Projects\github\jausions\TCPDF\include\barcodes\qrcode.php on line 895

Deprecated: Implicit conversion from float 18.5 to int loses precision in E:\Users\jausions\Documents\Projects\github\jausions\TCPDF\include\barcodes\qrcode.php on line 895

Deprecated: Implicit conversion from float 19.5 to int loses precision in E:\Users\jausions\Documents\Projects\github\jausions\TCPDF\include\barcodes\qrcode.php on line 895

Deprecated: Implicit conversion from float 20.5 to int loses precision in E:\Users\jausions\Documents\Projects\github\jausions\TCPDF\include\barcodes\qrcode.php on line 895

Deprecated: Implicit conversion from float 21.5 to int loses precision in E:\Users\jausions\Documents\Projects\github\jausions\TCPDF\include\barcodes\qrcode.php on line 895

Warning: Cannot modify header information - headers already sent by (output started at E:\Users\jausions\Documents\Projects\github\jausions\TCPDF\include\barcodes\qrcode.php:891) in E:\Users\jausions\Documents\Projects\github\jausions\TCPDF\tcpdf_barcodes_2d.php on line 90

Warning: Cannot modify header information - headers already sent by (output started at E:\Users\jausions\Documents\Projects\github\jausions\TCPDF\include\barcodes\qrcode.php:891) in E:\Users\jausions\Documents\Projects\github\jausions\TCPDF\tcpdf_barcodes_2d.php on line 91

Warning: Cannot modify header information - headers already sent by (output started at E:\Users\jausions\Documents\Projects\github\jausions\TCPDF\include\barcodes\qrcode.php:891) in E:\Users\jausions\Documents\Projects\github\jausions\TCPDF\tcpdf_barcodes_2d.php on line 92

Warning: Cannot modify header information - headers already sent by (output started at E:\Users\jausions\Documents\Projects\github\jausions\TCPDF\include\barcodes\qrcode.php:891) in E:\Users\jausions\Documents\Projects\github\jausions\TCPDF\tcpdf_barcodes_2d.php on line 93

Warning: Cannot modify header information - headers already sent by (output started at E:\Users\jausions\Documents\Projects\github\jausions\TCPDF\include\barcodes\qrcode.php:891) in E:\Users\jausions\Documents\Projects\github\jausions\TCPDF\tcpdf_barcodes_2d.php on line 94

Warning: Cannot modify header information - headers already sent by (output started at E:\Users\jausions\Documents\Projects\github\jausions\TCPDF\include\barcodes\qrcode.php:891) in E:\Users\jausions\Documents\Projects\github\jausions\TCPDF\tcpdf_barcodes_2d.php on line 95

---------------------------
File: barcodes/example_2d_qrcode_svgi.php
    Lint: PASSED
    Run: PASSED
    Output: NOT SVG FILE
    Logs:
---------------------------

Deprecated: Implicit conversion from float 0.5 to int loses precision in E:\Users\jausions\Documents\Projects\github\jausions\TCPDF\include\barcodes\qrcode.php on line 891

Deprecated: Implicit conversion from float 1.5 to int loses precision in E:\Users\jausions\Documents\Projects\github\jausions\TCPDF\include\barcodes\qrcode.php on line 891

Deprecated: Implicit conversion from float 2.5 to int loses precision in E:\Users\jausions\Documents\Projects\github\jausions\TCPDF\include\barcodes\qrcode.php on line 891

Deprecated: Implicit conversion from float 3.5 to int loses precision in E:\Users\jausions\Documents\Projects\github\jausions\TCPDF\include\barcodes\qrcode.php on line 891

Deprecated: Implicit conversion from float 4.5 to int loses precision in E:\Users\jausions\Documents\Projects\github\jausions\TCPDF\include\barcodes\qrcode.php on line 891

Deprecated: Implicit conversion from float 5.5 to int loses precision in E:\Users\jausions\Documents\Projects\github\jausions\TCPDF\include\barcodes\qrcode.php on line 891

Deprecated: Implicit conversion from float 6.5 to int loses precision in E:\Users\jausions\Documents\Projects\github\jausions\TCPDF\include\barcodes\qrcode.php on line 891

Deprecated: Implicit conversion from float 7.5 to int loses precision in E:\Users\jausions\Documents\Projects\github\jausions\TCPDF\include\barcodes\qrcode.php on line 891

Deprecated: Implicit conversion from float 8.5 to int loses precision in E:\Users\jausions\Documents\Projects\github\jausions\TCPDF\include\barcodes\qrcode.php on line 891

Deprecated: Implicit conversion from float 9.5 to int loses precision in E:\Users\jausions\Documents\Projects\github\jausions\TCPDF\include\barcodes\qrcode.php on line 891

Deprecated: Implicit conversion from float 10.5 to int loses precision in E:\Users\jausions\Documents\Projects\github\jausions\TCPDF\include\barcodes\qrcode.php on line 891

Deprecated: Implicit conversion from float 11.5 to int loses precision in E:\Users\jausions\Documents\Projects\github\jausions\TCPDF\include\barcodes\qrcode.php on line 891

Deprecated: Implicit conversion from float 12.5 to int loses precision in E:\Users\jausions\Documents\Projects\github\jausions\TCPDF\include\barcodes\qrcode.php on line 891

Deprecated: Implicit conversion from float 0.5 to int loses precision in E:\Users\jausions\Documents\Projects\github\jausions\TCPDF\include\barcodes\qrcode.php on line 895

Deprecated: Implicit conversion from float 1.5 to int loses precision in E:\Users\jausions\Documents\Projects\github\jausions\TCPDF\include\barcodes\qrcode.php on line 895

Deprecated: Implicit conversion from float 2.5 to int loses precision in E:\Users\jausions\Documents\Projects\github\jausions\TCPDF\include\barcodes\qrcode.php on line 895

Deprecated: Implicit conversion from float 3.5 to int loses precision in E:\Users\jausions\Documents\Projects\github\jausions\TCPDF\include\barcodes\qrcode.php on line 895

Deprecated: Implicit conversion from float 4.5 to int loses precision in E:\Users\jausions\Documents\Projects\github\jausions\TCPDF\include\barcodes\qrcode.php on line 895

Deprecated: Implicit conversion from float 5.5 to int loses precision in E:\Users\jausions\Documents\Projects\github\jausions\TCPDF\include\barcodes\qrcode.php on line 895

Deprecated: Implicit conversion from float 6.5 to int loses precision in E:\Users\jausions\Documents\Projects\github\jausions\TCPDF\include\barcodes\qrcode.php on line 895

Deprecated: Implicit conversion from float 7.5 to int loses precision in E:\Users\jausions\Documents\Projects\github\jausions\TCPDF\include\barcodes\qrcode.php on line 895

Deprecated: Implicit conversion from float 8.5 to int loses precision in E:\Users\jausions\Documents\Projects\github\jausions\TCPDF\include\barcodes\qrcode.php on line 895

Deprecated: Implicit conversion from float 9.5 to int loses precision in E:\Users\jausions\Documents\Projects\github\jausions\TCPDF\include\barcodes\qrcode.php on line 895

Deprecated: Implicit conversion from float 10.5 to int loses precision in E:\Users\jausions\Documents\Projects\github\jausions\TCPDF\include\barcodes\qrcode.php on line 895

Deprecated: Implicit conversion from float 11.5 to int loses precision in E:\Users\jausions\Documents\Projects\github\jausions\TCPDF\include\barcodes\qrcode.php on line 895

Deprecated: Implicit conversion from float 12.5 to int loses precision in E:\Users\jausions\Documents\Projects\github\jausions\TCPDF\include\barcodes\qrcode.php on line 895

Deprecated: Implicit conversion from float 13.5 to int loses precision in E:\Users\jausions\Documents\Projects\github\jausions\TCPDF\include\barcodes\qrcode.php on line 895

Deprecated: Implicit conversion from float 14.5 to int loses precision in E:\Users\jausions\Documents\Projects\github\jausions\TCPDF\include\barcodes\qrcode.php on line 895

Deprecated: Implicit conversion from float 15.5 to int loses precision in E:\Users\jausions\Documents\Projects\github\jausions\TCPDF\include\barcodes\qrcode.php on line 895

Deprecated: Implicit conversion from float 16.5 to int loses precision in E:\Users\jausions\Documents\Projects\github\jausions\TCPDF\include\barcodes\qrcode.php on line 895

Deprecated: Implicit conversion from float 17.5 to int loses precision in E:\Users\jausions\Documents\Projects\github\jausions\TCPDF\include\barcodes\qrcode.php on line 895

Deprecated: Implicit conversion from float 18.5 to int loses precision in E:\Users\jausions\Documents\Projects\github\jausions\TCPDF\include\barcodes\qrcode.php on line 895

Deprecated: Implicit conversion from float 19.5 to int loses precision in E:\Users\jausions\Documents\Projects\github\jausions\TCPDF\include\barcodes\qrcode.php on line 895

Deprecated: Implicit conversion from float 20.5 to int loses precision in E:\Users\jausions\Documents\Projects\github\jausions\TCPDF\include\barcodes\qrcode.php on line 895

Deprecated: Implicit conversion from float 21.5 to int loses precision in E:\Users\jausions\Documents\Projects\github\jausions\TCPDF\include\barcodes\qrcode.php on line 895

---------------------------
Test suite: FAILED
    Runtime: 19s
    Total tests: 83
    SKipped tests: 1
        example_006.php
    Failed tests: 4
        barcodes/example_2d_qrcode_html.php
        barcodes/example_2d_qrcode_png.php
        barcodes/example_2d_qrcode_svg.php
        barcodes/example_2d_qrcode_svgi.php
Exit code: 1
```

williamdes commented:
I think you should create a GitHub workflow so this tests run each time a commit is done.
Here is my tests: https://github.com/tecnickcom/TCPDF/blob/main/.github/workflows/tests.yml

Also If you can clarify that you are the author of some files it could be cool, maybe add an `@author` and `@licence`

jausions commented:
Interesting, I didn't know about it. I will look into it.

And, sure, I can add @author and @license.

jausions commented:
I've updated the commit. I just merely piggy-backed on your current test workflow for now. It would have been nice to include a run on Windows, but the sticky point is pdfinfo, although it's just a .zip to grab I think. 

An example of run can be seen here:

- https://github.com/jausions/TCPDF/actions/runs/3543554912

And we DO see PHP 8.1 failing (as well as PHP nightly), which I think is the most interesting part of the new test suite.

I also upgraded the actions since GitHub upgraded to Node16 and the previous ones were giving deprecation notices.
For instance, for the doc linter:

- https://github.com/jausions/TCPDF/actions/runs/3543554910

Finally, I added some `@author` and `@license` to my files.

fooman commented:
@jausions great work on the testsuite. I have also tried in the past a way to compare the generated pdfs against a baseline. While I didn't get to the bottom of it my working assumption was that it was due to the zip process not resulting in a deterministic result. Any pdfs that contain compressed elements would fail a bit by bit comparison as a result as well as included timestamps.

The approach I took in the end was to use Imagick to convert all pages of the pdf and then run an image comparison:
https://github.com/fooman/TCPDF/blob/main/tests/Common.php#L33
It's not perfect - I even had to start out with my own baseline pdfs as the pdfs I generated on my machine did not match the example pdfs on tcpdf.org (I blamed font and image rendering for this) but I think it's better than nothing. Let me know if you think there is merit in trying to integrate this approach with your PR.

jausions commented:
@williamdes I added Windows to the matrix for the test workflow. It seems to be stable and gives the same result as for Linux.

- https://github.com/jausions/TCPDF/actions/runs/3546050284

@fooman Yes, I did peer into the content of the PDF document itself, but too many small elements make it impossible to take a "dumb" approach. The conversion to image to compare might be an option indeed. As far as the baseline is concerned, I took the approach of comparing runs within the same environment to limit the risk of discrepancies due to OS-related setup. However, if all the fonts are provided, there should be no visual differences. Another bump in the road was for PNG images generated from the barcode examples. Somehow they were not the same size when changing PHP version, 7.2 being the threshold. A visual comparison might be doing the trick there too.

williamdes reviewed (approved):
💯

williamdes commented:
I like this, it's minimal and quite nice !
It's good that PHP 8.1 fails is there is still some fixes not related to this PR.

jausions commented:
I am still fine tuning the test suite, especially the command line options.

nicolaasuni commented on tests/vendor/README.md:
```diff
@@ -0,0 +1,5 @@
+Since Tecknicom/TCPDF works all the way back to PHP 5.3, it is not easy to have
```
I'd rather not include any third party software for licensing and legal reasons.
Instead we could just disable the tests for the PHP versions that are not supported by this.

jausions replied:
Sure, I'll look into it.

jausions replied:
Fixed.

williamdes commented on .github/workflows/tests.yml:
```diff
@@ -41,16 +44,20 @@ jobs:
                 flags: php-${{ matrix.php-version }}-${{ matrix.os }}
                 name: php-${{ matrix.php-version }}-${{ matrix.os }}
+            - name: Run PHP-based test suite
+              run: php ./tests/launch.php
```
You can extract `./tests/launch.sh` into a matrix variable like `php-version` so line `run: ./tests/launch.sh`
is `run: ${{ matrix.test-script }}`
That will allow you so add a bunch of `- { php-version: 'x.y', experimental: false, os: ubuntu-latest, coverage-extension: 'xdebug', test-script: './tests/launch.php' }`

And this will allow you to drop your old PHP compat code. It's okay if your test suite does not coverall old php versions

williamdes commented on .github/workflows/tests.yml:
```diff
@@ -15,42 +15,91 @@ jobs:
         strategy:
             fail-fast: false
             matrix:
+                #php-version: ["5.3", "5.4", "5.5", "5.6", "7.0", "7.1", "7.2", "7.3", "7.4", "8.0", "8.1", "8.2"]
```
Please add back newer php versions :)

jausions replied:
Yes, I will, I'm just trying to get things working on as many versions OS and PHP versions. It's an iterative process 😉

williamdes replied:
Very cool ;)

williamdes reviewed (dismissed):
Looks nice, is it ready for merge ?

williamdes commented:
Also could you check why codecov says there is a coverage decrease ?
Maybe we forget to upload the coverage ? Or to run the tests ?

jausions commented:
For instance this run: https://github.com/jausions/TCPDF/actions/runs/3679422459

I updated the workflow to keep the artifact in case of failure, so you will find the Zip file with it. However, I only put a retention of 1 day to not waste disk space.

The frustrating part for those 2 failing examples is no error and no PDF generated at all, both .error and .output files are empty.

So, for now, I disabled the PHP 7.0 run on Windows.
